### PR TITLE
feat: admin create faqs

### LINKED ISSUE #298: [FEAT]: Create an endpoint for Admin to Create FAQs
### Description
Create an endpoint that allows administrators to add new FAQs to the system. To help in managing frequently asked questions effectively and keeping the information up-to-date for users.
### Acceptance Criteria
- [ ] The endpoint /admin/faqs should accept a POST request.
- [ ] The request body must include question, answer, and createdBy.
- [ ] The request body may optionally include category and tags.
- [ ] Only authenticated administrators should be able to access this endpoint.
- [ ] On success, the endpoint should return a 201 Created status with the created FAQ object.
- [ ] If the input data is invalid, the endpoint should return a 400 Bad Request status with an appropriate error message.
- [ ] If the request is unauthorized, the endpoint should return a 401 Unauthorized status.
- [ ] In case of unexpected errors, the endpoint should return a 500 Internal Server Error status.
Endpoint:`api/v1/admin/faqs`
HTTP Method: POST
Request Body:
```json
{
  "question": "string",
  "answer": "string",
  "category": "string",
  "createdBy": "userId",
  "createdAt": "timestamp"
}
```
Response:
201 Created on success, returning the created FAQ object.
```json
{
  "success": true,
  "statuseCode": 201,
  "message": "string",
  "data": {
     "question": "string",
     "answer": "string",
     "category": "string",
     "createdBy": "userId",
     "createdAt": "timestamp"
   }
}
```
**400 Bad Request if the input data is invalid.**
```json
{
  "success": false,
  "message": "string",
  "errors": [
     "path": "string"
   ]
  "statusCode": 400
}
```
**401 Unauthorized if the request is not made by an admin.**
```json
{
  "success": false,
  "message": "string",
  "statusCode": 401
}
```
500 Internal Server Error for unexpected errors.
```json
{
  "success": false,
  "message": "string",
  "statusCode": 500
}
```
### Purpose
The purpose of this endpoint is to enable administrators to dynamically create and manage FAQs without requiring direct database access or manual updates. This will ensure that users always have access to the latest information.
### Requirements	
 - [ ] The request user must be authenticated with jwt token to retrieve the organisations
 - [ ] The request user must have admin role to access the endpoint
### Expected Outcome
- [ ] A new FAQ is created and stored in the system when valid data is provided by an authenticated admin.
- [ ] Appropriate error messages are returned for invalid input or unauthorized access.
- [ ] The new endpoint is covered by unit tests to ensure functionality and reliability.
### Unit Test
- [ ] Ensure the endpoint correctly identifies if the requesting user is signed in.
- [ ] Ensure the endpoint correctly identifies if the requesting user is an admin.
- [ ] Ensure the endpoint to handle all expected errors.
- [ ] Ensure the endpoint to successfully create a FAQs.

### PR DESCRIPTION
**Pull request by Demandtech**
close #298 

**what does this PR do**

This pr allow admin user to create FAQS

**How should this be manually tested?**

Authenticate a User:

Send a POST request to /api/v1/faqs to authenticate a user and obtain a JWT token.

Request Body
```
{
  "question": "string",
  "answer": "string",
  "category": "string",
  "tags:""
}
```
Note: category and tags are optional

Response
```
{
   "status": true,
   "status_code": 201,
   "message": "FAQ created successfully",
   "data":{
       "question": "wheres  you",
       "answer": "John",
       "category": "question",
       "tags": "CD",
       "id": "a9a8f1db-aedf-4863-9720-281f8b70cc70",
       "created_at": "2024-07-26T08:49:22.996Z",
       "updated_at": "2024-07-26T08:49:22.996Z"
   }
}
```

**403 Forbidden if the request is not made by an admin.**

```
{
     "status": false,
     "status_code": 403,
     "error":{
          "status": "Forbidden request",
          "message": "User not authorized to perform operation",
          "status_code": 403
     },
     "message": "User not authorized to perform operation"
}
```

**Invalid request**

```
{
      "status": false,
      "status_code": 400,
      "error": {
          "status": "Bad request",
          "message": "Faqs question exist",
          "status_code": 400
      }
     "message": "Faqs question exist"
}
```
**Checklist of what you did**

- [x]  The endpoint /admin/faqs should accept a POST request.
- [x]  The request body must include question,  answer.
- [x]  The request body may optionally include category and tags.
- [x]  Only authenticated administrators should be able to access this endpoint.
- [x]  On success, the endpoint should return a 201 Created status with the created FAQ object.
- [x]  If the input data is invalid, the endpoint should return a 400 Bad Request status with an appropriate error message.
- [x]  If the request is admin, the endpoint should return a 403 Unauthorized status.

**Test**
- [x]  Ensure the endpoint correctly identifies if the requesting user is signed in.
- [x]  Ensure the endpoint correctly identifies if the requesting user is an admin.
- [x]  Ensure the endpoint to handle all expected errors.
- [x]  Ensure the endpoint to successfully create a FAQs.

Test Screenshot
Screenshot 2024-07-26 at 10 10 54 AM

### CODE REVIEW

**Review by Heba-WebDev — APPROVED** [review]
Looks good to me.

File: src/modules/faqs/faqs.controller.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { Body, Controller, Post, Req, UseGuards } from '@nestjs/common';
+import { CreateFaqDto } from './dto/createFaqsDto';
+import { FaqsService } from './faqs.service';
+
+@Controller('admin/faqs')
+@UseGuards()
```
Comment (Homoakin619) [inline src/modules/faqs/faqs.controller.ts:6]: remove this

File: src/modules/faqs/faqs.controller.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+export class FaqsController {
+  constructor(private readonly faqService: FaqsService) {}
+  @Post('/')
+  async handleCreateFaqs(@Req() req, @Body() createFaqs: CreateFaqDto) {
+    const user = req['user'];
```
Comment (davidolanrewaju) [inline src/modules/faqs/faqs.controller.ts:11]: Document your route and remove the @UseGuards decorator

**Review by Heba-WebDev — CHANGES_REQUESTED** [review]
Update the title and format the first part of the description better

File: src/modules/faqs/faqs.service.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  constructor(
+    @InjectRepository(Faqs) private readonly faqsRepo: Repository<Faqs>,
+    @InjectRepository(User) private readonly userRepo: Repository<User>
+  ) {}
+
+  async createFaq(createFaqs: CreateFaqDto, userId: string) {
+    const user = await this.userRepo.findOneBy({ id: userId });
+
+    const existingFaq = await this.faqsRepo.findOneBy({ question: createFaqs.question });
+
```
Comment (davidolanrewaju) [inline src/modules/faqs/faqs.service.ts:19]: Wrap db calls in a try catch block

File: src/modules/faqs/faqs.controller.ts (hunk 2 above)
Comment (Demandtech) [reply src/modules/faqs/faqs.controller.ts:11]: I will do that now, thanks

File: src/modules/faqs/faqs.controller.ts (hunk 1 above)
Comment (Demandtech) [reply src/modules/faqs/faqs.controller.ts:6]: Okay boss

File: src/modules/faqs/faqs.service.ts (hunk 3 above)
Comment (Demandtech) [reply src/modules/faqs/faqs.service.ts:19]: Okay boss

**Comment by CFCIfe** [comment]
@Demandtech it should be 

> feat:
not
> Feat:

File: src/modules/faqs/dto/createFaqsDto.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { IsOptional, IsString } from 'class-validator';
+
+export class CreateFaqDto {
```
Comment (CFCIfe) [inline src/modules/faqs/dto/createFaqsDto.ts:3]: can we also add the `@IsNotEmpty()` class validator.

>@IsNotEmpty()	Checks if given value is not empty (!== '', !== null, !== undefined).

File: src/modules/faqs/faqs.service.ts [hunk 5]
```diff
@@ -13,9 +13,8 @@ export class FaqsService {
   ) {}
 
   async createFaq(createFaqs: CreateFaqDto, userId: string) {
-    const user = await this.userRepo.findOneBy({ id: userId });
-
     try {
+      const user = await this.userRepo.findOneBy({ id: userId });
       const existingFaq = await this.faqsRepo.findOneBy({ question: createFaqs.question });
```
Comment (CFCIfe) [inline src/modules/faqs/faqs.service.ts]: no catch block?

checkout the getUserRecord in the userService file. It handles this check.

Comment (Demandtech) [reply src/modules/faqs/faqs.service.ts]: There is a catch block, boss

File: src/modules/faqs/dto/createFaqsDto.ts (hunk 4 above)
Comment (Demandtech) [reply src/modules/faqs/dto/createFaqsDto.ts:3]: Alright boss

**Comment by Demandtech** [comment]
@CFCIfe @Homoakin619 @davidolanrewaju @Heba-WebDev  All requested changes have been made... Thanks

**Review by Heba-WebDev — DISMISSED** [review]

File: src/modules/faqs/faqs.service.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+@Injectable()
+export class FaqsService {
+  constructor(
+    @InjectRepository(Faqs) private readonly faqsRepo: Repository<Faqs>,
+    @InjectRepository(User) private readonly userRepo: Repository<User>
+  ) {}
+
+  async createFaq(createFaqs: CreateFaqDto, userId: string) {
+    try {
+      const user = await this.userRepo.findOneBy({ id: userId });
```
Comment (Homoakin619) [inline src/modules/faqs/faqs.service.ts:17]: Don't re invent the wheel. There is a method in the User service to get a user, use that

Comment (Demandtech) [reply src/modules/faqs/faqs.service.ts:17]: @Homoakin619 The methods are private for the user service, can I update it?

Comment (Heba-WebDev) [reply src/modules/faqs/faqs.service.ts:20]: @Demandtech You're only trying to find the user by id, nothing more that. So get it from the User service.

Comment (Homoakin619) [reply src/modules/faqs/faqs.service.ts:17]: there is a method in the UserService "getUserRecord" use that

Comment (Demandtech) [reply src/modules/faqs/faqs.service.ts:17]: But user_type was not added to the type interface... @Homoakin619

**Comment by Heba-WebDev** [comment]
New pr for this feature got merged today

**Pull request closed without merging**